components/style/selector_impl.rs

```diff
@@ -8,6 +8,73 @@ use selectors::parser::{ParserContext, SelectorImpl};
 pub enum PseudoElement {
     Before,
     After,
+    FirstLine,
+
+    // https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h
```

Comment:
Yes, but having a `PerPseudoElementSelectorMap` per possible pseudo seems overkill to me, but it might be doable, and it's, in fact, more ellegan than what we have right now.

I can try to do it first and get some numbers if you want.

I was talking about having two different `PseudoElement` types, as with two different `SelectorImpl`, but making a default impl feature gated, something like:

```rust
enum ServoPseudoElement {
    Before,
    After,
    FirstLine,
}

enum GeckoPseudoElement {
    Before,
    After,
    FirstLine,

     // ...
}

struct ServoSelectorImpl;
struct GeckoSelectorImpl;

impl ... for GeckoSelectorImpl { ... }
impl ... for ServoSelectorImpl { ... }

#[cfg(feature = "geckolib")]
pub type DefaultSelectorImpl = GeckoSelectorImpl;
#[cfg(not(feature = "geckolib"))]
pub type DefaultSelectorImpl = ServoSelectorImpl;
```

And we'll ve moving again with concrete types, and would work as long as both implementations have a `Before` and `After` variants.